pkg/hubble/observer/local_observer.go

```diff
@@ -681,3 +710,65 @@ func newRingReader(ring *container.Ring, req genericRequest, whitelist, blacklis
 	}
 	return container.NewRingReader(ring, idx), nil
 }
+
+type namespaceRecord struct {
+	namespace *observerpb.Namespace
+	added     time.Time
+}
+
+type namespaceManager struct {
+	mu         lock.RWMutex
+	namespaces map[string]namespaceRecord
+}
+
+func NewNamespaceManager() *namespaceManager {
+	return &namespaceManager{
+		namespaces: make(map[string]namespaceRecord),
+	}
+}
+
+func (m *namespaceManager) Run(ctx context.Context) {
+	ticker := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case t := <-ticker.C:
+			// periodically remove any namespaces which haven't been seen in flows
+			// for the last hour
+			m.mu.Lock()
+			for key, record := range m.namespaces {
+				if record.added.Add(time.Hour).After(t) {
+					delete(m.namespaces, key)
+				}
+			}
+			m.mu.Unlock()
+		}
+	}
+}
+
+func (m *namespaceManager) GetNamespaces() []*observerpb.Namespace {
+	m.mu.RLock()
+	namespaces := make([]*observerpb.Namespace, 0, len(m.namespaces))
+	for _, ns := range m.namespaces {
+		namespaces = append(namespaces, ns.namespace)
+	}
+	m.mu.RUnlock()
+
+	slices.SortFunc(namespaces, func(a, b *observerpb.Namespace) bool {
+		if a.Cluster != b.Cluster {
+			return a.Cluster < b.Cluster
+		}
+		return a.Namespace < b.Namespace
+	})
+	return namespaces
+}
+
+func (m *namespaceManager) AddNamespace(ns *observerpb.Namespace) {
+	m.mu.Lock()
+	key := ns.GetCluster() + "/" + ns.GetNamespace()
+	m.namespaces[key] = namespaceRecord{namespace: ns, added: time.Now()}
```

Comment:
We could add `now` as param for testing here too, and callers would use `NowFunc`

```suggestion
func (m *namespaceManager) AddNamespace(now time.Time, ns *observerpb.Namespace) {
	m.mu.Lock()
	key := ns.GetCluster() + "/" + ns.GetNamespace()
	m.namespaces[key] = namespaceRecord{namespace: ns, added: now}
```

Reply:
I think having `now` as a param is a bit of a leaky abstraction, instead we can make it a field on namespaceManager for testing. We might not use time in a different implementation of a NamespaceManager, so I think this would be  more flexible.